### PR TITLE
Split debugging into a separate library

### LINKED ISSUE #91: Split debug library from Foundation
Split the low-level debugging utilities out of foundation and into their own library. This would include the callstack utilities and assertion dialog. It would also be the home for any newer low-level debugging components, like a memory debugger or the like.

This is a required step to making foundation an interface library.

### PR DESCRIPTION
**Pull request by seanmiddleditch**
Foundation is now an interface-only library, making it easier to use from anywhere. It does still have a dependency on the new runtime library, though, for assertion handling; a better way to handle this might be good.

Fixes #91 and #92

### CODE REVIEW

File: source/library/ecs/private/world.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // Copyright (C) 2019 Sean Middleditch, all rights reserverd.
 
+#include <potato/runtime/assertion.h>
```
Comment (marcinwie) [inline source/library/ecs/private/world.cpp]: starting to mix < with " around headers?

File: source/library/foundation/CMakeLists.txt [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
 add_library(potato::foundation ALIAS foundation)
 
-include(up_set_common_properties)
-up_set_common_properties(foundation)
+#include(up_set_common_properties)
+#up_set_common_properties(foundation)
 
 target_link_libraries(foundation
-    PUBLIC
+    INTERFACE
```
[the remark this answers is not in the document]
Comment (seanmiddleditch) [reply source/library/foundation/CMakeLists.txt]: Open to discussion on that. I think if it's suitable low-level and important, sure. I wouldn't want it to be a dumping ground for random shit, though.  Can always add more libraries easily enough. :)

File: source/library/ecs/private/world.cpp (hunk 1 above)
Comment (seanmiddleditch) [reply source/library/ecs/private/world.cpp]: I keep going back and forth on them. Should pick a standard, document it, and figure out how to make clang-tidy enforce it.

Given both, what do you think? I'm a little partial to `<>` and starting to lean towards the idea that `""` should be limited for local header inclusion only (and maybe not even then).

But that's just an opinion and obviously I waffle on it sometimes. :)

Comment (marcinwie) [reply source/library/ecs/private/world.cpp]: i'm used to the idea that all 'external' headers use <> and local use "" -- somehow it makes a difference for me -- i know compiler doesn't care.  

speaking off -- i'm not a big fan of potato prefix in paths -- feels like we should be able to configure everything so that library name is the first name in include.. so in this case it would be 

`#include <runtime/assertion.h>` 

long term it makes it easier to move libraries around without having to change bunch of code.

Comment (seanmiddleditch) [reply source/library/ecs/private/world.cpp]: @marcinwie:

> speaking off -- i'm not a big fan of potato prefix in paths -- feels like we should be able to configure everything so that library name is the first name in include.. so in this case it would be
>
> `#include <runtime/assertion.h>`
>
> long term it makes it easier to move libraries around without having to change bunch of code.

I don't follow the argument that removing `potato/` makes any difference in moving code around, as the `potato/` part will be left alone.

There's also a strong case to be made about namespace headers, too. Names like `runtime` are not exactly unique, but `potato/runtime` is quite unique. As the project grows, it'll help differentiate headers that are from Potato libraries vs those that are from other dependencies that might have non-unique "namespaces." You effectively either way weird unique names like `formatxx/` or you want a namespace like `seanmiddleditch/format` but you'd not want a generic name like `format`.

tl;dr - the `potato/` prefix makes it abundantly clear that a header is one of _ours_ and I don't see the cost that you're seeing.

> i'm partial to the idea that all 'external' headers use <> and local use "" -- somehow it makes a difference for me -- i know compiler doesn't care.

I can certainly follow that argument. The debate then is whether other Potato libraries are "external" or not. :)

I guess part of what I think is that we might want to start factoring out the codebase eventually into reusable standalone libraries. e.g., if we do the `foundation` -> `spud` thing I was talking about, I'd move `spud` into its own repo and make it independent of Potato to the extent reasonable (probably keep the `up::` C++ namespace for our convenience, but the header paths could reasonably change from `potato/foundation/foo.h` to just `spud/foo.h`, I think).
